File: scripts/approve-pr-when-clean.mjs
#!/usr/bin/env node

import { pathToFileURL } from "node:url";
import { githubRequest } from "./shared/github-client.mjs";

function normalizeText(value) {
  return String(value ?? "")
    .replace(/\r\n/g, "\n")
    .trim();
}

function parseRepository(repo) {
  const [owner, name] = String(repo || "").split("/");
  if (!owner || !name) {
    throw new Error("GITHUB_REPOSITORY must be owner/name");
  }
  return { owner, name };
}

function parsePullNumber(value) {
  const parsed = Number.parseInt(String(value || "").trim(), 10);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    throw new Error("PR_NUMBER must be a positive integer");
  }
  return parsed;
}

function isPermissionIssue(error) {
  const message = normalizeText(error?.message || "").toLowerCase();
  return (
    message.includes("resource not accessible by integration")
    || message.includes("not permitted to create or approve pull requests")
    || message.includes("must have write access")
  );
}

async function main() {
  await approvePrWhenClean({
    token: process.env.GITHUB_TOKEN || "",
    repo: process.env.GITHUB_REPOSITORY || "",
    prNumber: process.env.PR_NUMBER || "",
    expectedHeadSha: process.env.SHA || "",
  });
}

export async function approvePrWhenClean({
  token,
  repo,
  prNumber,
  expectedHeadSha,
  request = githubRequest,
  stderr = process.stderr,
}) {
  const normalizedToken = normalizeText(token || "");
  const normalizedRepo = normalizeText(repo || "");
  const normalizedExpectedHeadSha = normalizeText(expectedHeadSha || "");
  const normalizedPrNumber = parsePullNumber(prNumber || "");
  if (!normalizedToken) {
    throw new Error("GITHUB_TOKEN is required");
  }

  const { owner, name } = parseRepository(normalizedRepo);
  const pull = await request({
    method: "GET",
    token: normalizedToken,
    url: `https://api.github.com/repos/${owner}/${name}/pulls/${normalizedPrNumber}`,
  });

  const currentHeadSha = normalizeText(pull?.head?.sha);
  if (normalizedExpectedHeadSha && currentHeadSha && normalizedExpectedHeadSha !== currentHeadSha) {
    stderr.write(
      `[approve-pr-when-clean] skipped: PR head moved from ${normalizedExpectedHeadSha} to ${currentHeadSha}\n`,
    );
    return;
  }

  const body = "LGTM automation: no open findings in the latest run.";
  try {
    await request({
      method: "POST",
      token: normalizedToken,
      url: `https://api.github.com/repos/${owner}/${name}/pulls/${normalizedPrNumber}/reviews`,
      body: {
        event: "APPROVE",
        body,
      },
    });
  } catch (error) {
    if (isPermissionIssue(error)) {
      stderr.write(
        `[approve-pr-when-clean] non-fatal: unable to auto-approve (${normalizeText(error?.message)})\n`,
      );
      return;
    }
    throw error;
  }
}

function isCliMain() {
  if (!process.argv[1]) return false;
  return import.meta.url === pathToFileURL(process.argv[1]).href;
}

if (isCliMain()) {
  main().catch((error) => {
    console.error(error.message);
    process.exit(1);
  });
}

File: scripts/approve_pr_when_clean.py
from __future__ import annotations

import os
import sys
from typing import Protocol

from scripts.shared.github_client import github_request
from scripts.shared.types import JSONObject


class RequestFunc(Protocol):
    def __call__(
        self,
        *,
        method: str,
        url: str,
        token: str,
        body: JSONObject | None = None,
    ) -> JSONObject | list[JSONObject] | str | int | float | bool | None: ...


def normalize_text(value: str | int | float | bool | None) -> str:
    return str("" if value is None else value).replace("\r\n", "\n").strip()


def parse_repository(repo: str) -> tuple[str, str]:
    parts = repo.split("/")
    if len(parts) != 2 or not parts[0] or not parts[1]:
        raise ValueError("GITHUB_REPOSITORY must be owner/name")
    return parts[0], parts[1]


def parse_pull_number(value: str) -> int:
    try:
        parsed = int(value.strip())
    except ValueError as error:
        raise ValueError("PR_NUMBER must be a positive integer") from error
    if parsed <= 0:
        raise ValueError("PR_NUMBER must be a positive integer")
    return parsed


def is_permission_issue(error_message: str) -> bool:
    message = normalize_text(error_message).lower()
    return (
        "resource not accessible by integration" in message
        or "not permitted to create or approve pull requests" in message
        or "must have write access" in message
    )


def approve_pr_when_clean(
    *,
    token: str,
    repo: str,
    pr_number: str,
    expected_head_sha: str,
    request: RequestFunc = github_request,
) -> None:
    normalized_token = normalize_text(token)
    normalized_repo = normalize_text(repo)
    normalized_expected_head_sha = normalize_text(expected_head_sha)
    normalized_pr_number = parse_pull_number(pr_number)

    if not normalized_token:
        raise ValueError("GITHUB_TOKEN is required")

    owner, name = parse_repository(normalized_repo)
    pull = request(
        method="GET",
        token=normalized_token,
        url=f"https://api.github.com/repos/{owner}/{name}/pulls/{normalized_pr_number}",
    )
    if not isinstance(pull, dict):
        raise ValueError("Unable to load pull request payload")

    current_head_sha = ""
    head_payload = pull.get("head")
    if isinstance(head_payload, dict):
        sha_value = head_payload.get("sha")
        if isinstance(sha_value, str):
            current_head_sha = normalize_text(sha_value)

    if normalized_expected_head_sha and current_head_sha and normalized_expected_head_sha != current_head_sha:
        sys.stderr.write(
            f"[approve-pr-when-clean] skipped: PR head moved from {normalized_expected_head_sha} to {current_head_sha}\n"
        )
        return

    body = "LGTM automation: no open findings in the latest run."
    try:
        request(
            method="POST",
            token=normalized_token,
            url=f"https://api.github.com/repos/{owner}/{name}/pulls/{normalized_pr_number}/reviews",
            body={"event": "APPROVE", "body": body},
        )
    except Exception as error:
        message = str(error)
        if is_permission_issue(message):
            sys.stderr.write(
                f"[approve-pr-when-clean] non-fatal: unable to auto-approve ({normalize_text(message)})\n"
            )
            return
        raise


def main() -> None:
    approve_pr_when_clean(
        token=os.getenv("GITHUB_TOKEN", ""),
        repo=os.getenv("GITHUB_REPOSITORY", ""),
        pr_number=os.getenv("PR_NUMBER", ""),
        expected_head_sha=os.getenv("SHA", ""),
    )


if __name__ == "__main__":
    main()
